### PR TITLE
Add @overload decorator and accessor(s)

### PR DESCRIPTION
**Pull request by johanste**
New overload decorator along with accessor to see what overloads exist for a given operation as well as which operation (if any) a given operation overloads.

No emitters were updated to take advantage of the new decorator.

### CODE REVIEW

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [inline packages/compiler/lib/decorators.ts:826]: Should this be `overloadOf` so that it reads more clearly?  I don't have a strong opinion here.

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [inline packages/compiler/lib/decorators.ts:869]: Might call this `isOverload`

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [inline packages/compiler/lib/decorators.ts:856]: Might call this `getOverloads`

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [inline packages/compiler/lib/decorators.ts:838]: You won't need a custom diagnostic if you use `createDecoratorDefinition`

**Review by daviwil — APPROVED** [review]
Overall this looks good to me barring some possible naming changes.  Might be good for @bterlson or @nguerrera to chime in on scenarios where overloads can be defined, though it's probably not very controversial.

File: packages/compiler/lib/decorators.ts
Comment (nguerrera) [reply packages/compiler/lib/decorators.ts:826]: I recall people liking `@overloads(generalOp)` as in "it overloads generalOp", but I'm fine with `@overload` or `@overloadOf` too.

File: packages/compiler/lib/decorators.ts
Comment (nguerrera) [inline packages/compiler/lib/decorators.ts:832]: I'd be really curious how easy this is with `checker.isTypeAssignableTo`. Unless it's especially hard (which I hope it isn't), then I worry about leaving this TODO tool long and then having to make a breaking change.

File: packages/compiler/lib/decorators.ts
Comment (johanste) [reply packages/compiler/lib/decorators.ts:869]: I originally had the `isOverload` and `getOverload` and I found myself mistaking them for each other - e.g. how is `isOverload()` different than `getOverloads() !== undefined`. I don't have a strong opinion on the names, though.

File: packages/compiler/lib/decorators.ts
Comment (johanste) [reply packages/compiler/lib/decorators.ts:826]: I mean, it is not entirely impossible that I had inspiration from [elsewhere](https://docs.python.org/3/library/typing.html#typing.overload) when I picked the name. The main issue I have with the trailing `s` is that it becomes a plural-s.

File: packages/compiler/lib/decorators.ts
Comment (johanste) [reply packages/compiler/lib/decorators.ts:832]: I just didn't know which method to use :). I have no problems other than TLBV (time left before vacation) to make this change before merging.

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [reply packages/compiler/lib/decorators.ts:869]: I think it's pretty common in Cadl decorators to have both an `isThing` and `getThings` so it wouldn't be so out of place.

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [reply packages/compiler/lib/decorators.ts:826]: Prior art is good to see, I don't think the slight differences in name will make a huge difference either way so we can keep it as it is if you don't prefer any of the other options.

File: packages/compiler/lib/decorators.ts
Comment (johanste) [reply packages/compiler/lib/decorators.ts:869]: I guess the verb/noun duality of `overload` is what trips me (as a non-native English speaker) up (and makes the trailing 's' ambiguous - overloads as a noun is plural, as a verb it is in third person singular. Or whatever it is called in English). I will defer to people who speak better English than me.

File: packages/compiler/lib/decorators.ts
Comment (markcowl) [reply packages/compiler/lib/decorators.ts:856]: getOverloadedBy [executed against the overloaded method]
getOverloads

isOverloading [executed against the overloading method]
isOverload
getOverloadTarget

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [reply packages/compiler/lib/decorators.ts]: Works pretty well!  Added usage of `isTypeAssignableTo` and a test to verify.

File: packages/compiler/lib/decorators.ts
Comment (daviwil) [reply packages/compiler/lib/decorators.ts]: Went with `getOverloads` and `getOverloadedOperation`.

**Pull request merged**
